### ISSUE TITLE
Do activity in one of a pair of concurrent states is not being run

### ISSUE THREAD

**Issue by Nava2** [umple/umple#207]
Originally reported on Google Code with ID 307
Owned by `gabbourget`

---

**What steps will reproduce the problem?**

Enter the code:

class X {
  sm {
    s1 {
      entry / {z();}
      -> s2;
    }
    s2 {
      s2a {
        do {longlastingmethod(); }
      }
      ||
      s2b {
        e2 -> s1;
      }
    }
  }
}

**What is the expected output? What do you see instead?**

longlastingmethod() should be run when we enter s2a, but it is not happening

---

Reported by @umple on 2012-06-13 17:37:22

**Comment by Nava2** [comment]
Looks like more generally, it's nested state machine, not specifically  concurrent
states:

class X {
  sm {
    s2 {
      s2a {
        do {longlastingmethod(); }
      }
    }
  }
}

---

Reported by @aforward on 2012-06-13 23:24:41

**Comment by Nava2** [comment]
Reported by @umple on 2012-06-14 18:03:32
- **Labels added**: Priority-High, Diffic-Med
- **Labels removed**: Priority-Medium

**Comment by Nava2** [comment]
Raising this to a critical issue. I need this working for the demo in Montreal in December.
I just spend hours working on an example that does not work. I am really annoyed it
has been sitting for so long. Do activities do not run at all when we have nested state
machines. This is a huge failure case.

---

Reported by @umple on 2012-11-22 20:32:51
- **Labels added**: Priority-Critical
- **Labels removed**: Priority-High

**Comment by Nava2** [comment]
Reported by @umple on 2013-01-19 15:12:59
- **Owner removed**
- **Labels added**: ucosp
- **cc added:** +@obahy

**Comment by Nava2** [comment]
Reported by @umple on 2013-01-19 18:24:06
- **Owner changed**: `gabbourget`

**Comment by Nava2** [comment]
The problem was in the code generation. It was looping in all statemachines, but not
in the nested statemachines. I just added the loop in the JET files so it generates
the activities and classes extending Thread. The code could need a refactor since it
contains a duplication of code.

I attached the patch to this comment.

---

Reported by `gabbourget` on 2013-02-06 20:57:20
- **cc added:** +@umple
- **cc removed:** -@obahy
  - _Attachment: [patch-GABRIELBOURGET-307-activityInNestedStateMachine.diff](https://storage.googleapis.com/google-code-attachments/umple/issue-307/comment-7/patch-GABRIELBOURGET-307-activityInNestedStateMachine.diff)_
